Rebase on Go 1.19.5

Lekensteyn opened the pull request:
Locally verified with:
```bash
git rebase --exec "bin/go test -run 'Test[^P].*|TestP[^l].*|TestPl[^a].*' ./src/crypto/tls ./src/crypto/x509" go1.19.5
```

Changes to resolve conflicts and compile errors:
- commit afc5595d58 (add CIRCL sigalgs): Rename `supportedSignatureAlgorithms` to `defaultSupportedSignatureAlgorithms` (it was renamed in Go 1.19 due to boringcrypto).
- commit 8c79656aef (add PQSignatureSchemesEnabled option):  incorporate `supportedSignatureAlgorithms()` boringcrypto compatibility in the internal `tls.Config#supportedSignatureAlgorithms` function.
- commit f1691ca795 (add DC): (trivial) sort `AllowDC` in the reindented documentation comment.
- commit 7fb630ea4c (add ECH): rewrite `supportedVersionsFromMin` to enable code reuse, it failed to compile due to removal of `debugEnableTLS10`

<details>
<summary>git range-diff go1.18.8..cf go1.19.5..pwu/cf-1.19.5</summary>

```
 1:  1bc349f07b <  -:  ---------- VERSION: add cf tag
 -:  ---------- >  1:  81144ed398 VERSION: add cf tag
 2:  dc7af31443 =  2:  96ba4f2dcc cmd/go/internal/imports, go/build: add cfgo build tag
 3:  9ca0466b24 =  3:  f3b0f0f926 .github: add docker-compose.yaml and GitHub Action for CI
 4:  1e76e0d04c =  4:  89a9569c58 .github: replace upstream templates and CODE_OF_CONDUCT.md
 5:  898fb03c08 !  5:  c4feb39122 cmd/api, go/build: ignore CF-specific API changes
    @@ src/cmd/api/goapi.go: func main() {
      	}
     
      ## src/cmd/api/run.go ##
    -@@ src/cmd/api/run.go: func allowNew(apiDir string) string {
    +@@ src/cmd/api/run.go: func allowNew() string {
      		log.Fatalln("Problem with api file for previous release:", err)
      	}
      
    @@ src/cmd/api/run.go: func allowNew(apiDir string) string {
     +
      	// See whether the api/go1.n.txt for this Go version has been created.
      	// (As of April 2021, it gets created during the release of the first Beta.)
    - 	_, err := os.Stat(filepath.Join(apiDir, fmt.Sprintf("go1.%d.txt", goversion.Version)))
    + 	_, err := os.Stat(fmt.Sprintf("go1.%d.txt", goversion.Version))
     
      ## src/go/build/deps_test.go ##
     @@ src/go/build/deps_test.go: func TestDependencies(t *testing.T) {
 6:  b4e5b3faaf =  6:  6a873c45b0 circl: import crypto library CIRCL
 7:  b7239ebf48 !  7:  afc5595d58 crypto: add support for CIRCL signature schemes
    @@ src/crypto/tls/tls_cf.go (new)
     +	// Note: this will extend the signature_algorithms TLS extension in the
     +	// Client Hello which requires changes to various testdata files.
     +	for _, cs := range circlSchemes {
    -+		supportedSignatureAlgorithms = append(supportedSignatureAlgorithms,
    ++		defaultSupportedSignatureAlgorithms = append(defaultSupportedSignatureAlgorithms,
     +			SignatureScheme(cs.scheme.(circlPki.TLSScheme).TLSIdentifier()))
     +	}
     +}
 8:  ce359c26cf !  8:  8c79656aef crypto/tls: add PQSignatureSchemesEnabled option
    @@ Commit message
     
             bin/go run src/crypto/tls/generate_cert.go -host example.golang -circl Ed25519-Dilithium3 -duration 87600h
     
    +    Since Go 1.19, FIPS-only mode must remain disabled to enable PQ sigalgs.
    +
      ## src/crypto/tls/common.go ##
     @@ src/crypto/tls/common.go: type Config struct {
      	// its key share in TLS 1.3. This may change in the future.
    @@ src/crypto/tls/handshake_client.go: func (c *Conn) makeClientHello() (*clientHel
      	}
      
      	if hello.vers >= VersionTLS12 {
    --		hello.supportedSignatureAlgorithms = supportedSignatureAlgorithms
    +-		hello.supportedSignatureAlgorithms = supportedSignatureAlgorithms()
     +		hello.supportedSignatureAlgorithms = config.supportedSignatureAlgorithms()
      	}
    - 
    - 	var params ecdheParameters
    + 	if testingOnlyForceClientHelloSignatureAlgorithms != nil {
    + 		hello.supportedSignatureAlgorithms = testingOnlyForceClientHelloSignatureAlgorithms
     
      ## src/crypto/tls/handshake_client_tls13.go ##
     @@ src/crypto/tls/handshake_client_tls13.go: func (hs *clientHandshakeStateTLS13) readServerCertificate() error {
      	}
      
      	// See RFC 8446, Section 4.4.3.
    --	if !isSupportedSignatureAlgorithm(certVerify.signatureAlgorithm, supportedSignatureAlgorithms) {
    +-	if !isSupportedSignatureAlgorithm(certVerify.signatureAlgorithm, supportedSignatureAlgorithms()) {
     +	if !isSupportedSignatureAlgorithm(certVerify.signatureAlgorithm, c.config.supportedSignatureAlgorithms()) {
      		c.sendAlert(alertIllegalParameter)
      		return errors.New("tls: certificate used with invalid signature algorithm")
    @@ src/crypto/tls/handshake_server.go: func (hs *serverHandshakeState) doFullHandsh
      		}
      		if c.vers >= VersionTLS12 {
      			certReq.hasSignatureAlgorithm = true
    --			certReq.supportedSignatureAlgorithms = supportedSignatureAlgorithms
    +-			certReq.supportedSignatureAlgorithms = supportedSignatureAlgorithms()
     +			certReq.supportedSignatureAlgorithms = c.config.supportedSignatureAlgorithms()
      		}
      
    @@ src/crypto/tls/handshake_server_tls13.go: func (hs *serverHandshakeStateTLS13) s
      		certReq := new(certificateRequestMsgTLS13)
      		certReq.ocspStapling = true
      		certReq.scts = true
    --		certReq.supportedSignatureAlgorithms = supportedSignatureAlgorithms
    +-		certReq.supportedSignatureAlgorithms = supportedSignatureAlgorithms()
     +		certReq.supportedSignatureAlgorithms = c.config.supportedSignatureAlgorithms()
      		if c.config.ClientCAs != nil {
      			certReq.certificateAuthorities = c.config.ClientCAs.Subjects()
    @@ src/crypto/tls/handshake_server_tls13.go: func (hs *serverHandshakeStateTLS13) r
      		}
      
      		// See RFC 8446, Section 4.4.3.
    --		if !isSupportedSignatureAlgorithm(certVerify.signatureAlgorithm, supportedSignatureAlgorithms) {
    +-		if !isSupportedSignatureAlgorithm(certVerify.signatureAlgorithm, supportedSignatureAlgorithms()) {
     +		if !isSupportedSignatureAlgorithm(certVerify.signatureAlgorithm, c.config.supportedSignatureAlgorithms()) {
      			c.sendAlert(alertIllegalParameter)
      			return errors.New("tls: client certificate used with invalid signature algorithm")
    @@ src/crypto/tls/tls_cf.go: func sigTypeByCirclScheme(scheme circlSign.Scheme) uin
     +var supportedSignatureAlgorithmsWithCircl []SignatureScheme
     +
     +// supportedSignatureAlgorithms returns enabled signature schemes. PQ signature
    -+// schemes are only included when tls.Config#PQSignatureSchemesEnabled is set.
    ++// schemes are only included when tls.Config#PQSignatureSchemesEnabled is set
    ++// and FIPS-only mode is not enabled.
     +func (c *Config) supportedSignatureAlgorithms() []SignatureScheme {
    ++	// If FIPS-only mode is requested, do not add other algos.
    ++	if needFIPS() {
    ++		return supportedSignatureAlgorithms()
    ++	}
     +	if c != nil && c.PQSignatureSchemesEnabled {
     +		return supportedSignatureAlgorithmsWithCircl
      	}
     -	// Note: this will extend the signature_algorithms TLS extension in the
     -	// Client Hello which requires changes to various testdata files.
    -+	return supportedSignatureAlgorithms
    ++	return defaultSupportedSignatureAlgorithms
     +}
     +
     +func init() {
    -+	supportedSignatureAlgorithmsWithCircl = append([]SignatureScheme{}, supportedSignatureAlgorithms...)
    ++	supportedSignatureAlgorithmsWithCircl = append([]SignatureScheme{}, defaultSupportedSignatureAlgorithms...)
      	for _, cs := range circlSchemes {
    --		supportedSignatureAlgorithms = append(supportedSignatureAlgorithms,
    +-		defaultSupportedSignatureAlgorithms = append(defaultSupportedSignatureAlgorithms,
     +		supportedSignatureAlgorithmsWithCircl = append(supportedSignatureAlgorithmsWithCircl,
      			SignatureScheme(cs.scheme.(circlPki.TLSScheme).TLSIdentifier()))
      	}
 9:  e7294ebb29 =  9:  de9cdcf3c9 crypto/tls: add CFControl parameter to Config
10:  7286b3aeab = 10:  2814068d4e crypto/tls: add Config.CFEventHandler to expose intra-handshake timing
11:  cf7f10408d ! 11:  f1691ca795 crypto/tls: implement draft-ietf-tls-subcerts-10
    @@ Commit message
         Fixes issues in earlier patch, addressing #127, #128, #129, #130, and #131.
         Add tool for generating delegated credentials.
     
    -    Co-authored-by: jhoyla <jhoyla@users.noreply.github.com>
    +    Co-authored-by: Jonathan Hoyland <jhoyland@cloudflare.com>
     
      ## src/crypto/tls/auth.go ##
     @@ src/crypto/tls/auth.go: func signatureSchemesForCertificate(version uint16, cert *Certificate) []Signatu
    @@ src/crypto/tls/common.go: const (
      	extensionSessionTicket           uint16 = 35
      	extensionPreSharedKey            uint16 = 41
      	extensionEarlyData               uint16 = 42
    -@@ src/crypto/tls/common.go: var supportedSignatureAlgorithms = []SignatureScheme{
    +@@ src/crypto/tls/common.go: var defaultSupportedSignatureAlgorithms = []SignatureScheme{
      	ECDSAWithSHA1,
      }
      
    @@ src/crypto/tls/handshake_messages.go: func unmarshalCertificate(s *cryptobyte.St
      ## src/crypto/tls/handshake_messages_test.go ##
     @@ src/crypto/tls/handshake_messages_test.go: func (*clientHelloMsg) Generate(rand *rand.Rand, size int) reflect.Value {
      	if rand.Intn(10) > 5 {
    - 		m.supportedSignatureAlgorithmsCert = supportedSignatureAlgorithms
    + 		m.supportedSignatureAlgorithmsCert = supportedSignatureAlgorithms()
      	}
     +	if rand.Intn(10) > 5 {
     +		m.delegatedCredentialSupported = true
    @@ src/crypto/tls/handshake_messages_test.go: func (*certificateRequestMsgTLS13) Ge
     +		m.supportedSignatureAlgorithmsDC = supportedSignatureAlgorithmsDC
     +	}
      	if rand.Intn(10) > 5 {
    - 		m.supportedSignatureAlgorithms = supportedSignatureAlgorithms
    + 		m.supportedSignatureAlgorithms = supportedSignatureAlgorithms()
      	}
     @@ src/crypto/tls/handshake_messages_test.go: func (*certificateMsgTLS13) Generate(rand *rand.Rand, size int) reflect.Value {
      		m.ocspStapling = true
    @@ src/crypto/x509/x509.go: func buildCertExtensions(template *Certificate, subject
      		ret[n].Id = oidExtensionAuthorityKeyId
      		ret[n].Value, err = asn1.Marshal(authKeyId{authorityKeyId})
     @@ src/crypto/x509/x509.go: var emptyASN1Subject = []byte{0x30, 0}
    - //  - ExtraExtensions
    - //  - IPAddresses
    - //  - IsCA
    -+//  - AllowDC
    - //  - IssuingCertificateURL
    - //  - KeyUsage
    - //  - MaxPathLen
    + // CreateCertificate creates a new X.509 v3 certificate based on a template.
    + // The following members of template are currently used:
    + //
    ++//   - AllowDC
    + //   - AuthorityKeyId
    + //   - BasicConstraintsValid
    + //   - CRLDistributionPoints
     
      ## src/crypto/x509/x509_test.go ##
     @@ src/crypto/x509/x509_test.go: func TestISOOIDInCertificate(t *testing.T) {
12:  8e8e677876 ! 12:  7fb630ea4c crypto/tls: implement draft-ietf-tls-esni-13
    @@ src/crypto/tls/common.go: func (c *Config) supportedVersions(isClient bool) []ui
      }
      
     +func (c *Config) supportedVersionsFromMin(isClient bool, minVersion uint16) []uint16 {
    -+	versions := make([]uint16, 0, len(supportedVersions))
    -+	for _, v := range supportedVersions {
    -+		if (c == nil || c.MinVersion == 0) && !debugEnableTLS10 &&
    -+			isClient && v < VersionTLS12 {
    -+			continue
    ++	versions := c.supportedVersions(isClient)
    ++	filteredVersions := versions[:0]
    ++	for _, v := range versions {
    ++		if v >= minVersion {
    ++			filteredVersions = append(filteredVersions, v)
     +		}
    -+		if c != nil && c.MinVersion != 0 && v < c.MinVersion {
    -+			continue
    -+		}
    -+		if c != nil && c.MaxVersion != 0 && v > c.MaxVersion {
    -+			continue
    -+		}
    -+		if v < minVersion {
    -+			continue
    -+		}
    -+		versions = append(versions, v)
     +	}
    -+	return versions
    ++	return filteredVersions
     +}
     +
      func (c *Config) maxSupportedVersion(isClient bool) uint16 {
    @@ src/crypto/tls/ech_test.go (new)
     
      ## src/crypto/tls/handshake_client.go ##
     @@ src/crypto/tls/handshake_client.go: type clientHandshakeState struct {
    - 	session      *ClientSessionState
    - }
    + 
    + var testingOnlyForceClientHelloSignatureAlgorithms []SignatureScheme
      
     -func (c *Conn) makeClientHello() (*clientHelloMsg, ecdheParameters, error) {
     +func (c *Conn) makeClientHello(minVersion uint16) (*clientHelloMsg, ecdheParameters, error) {
    @@ src/crypto/tls/handshake_client.go: func (c *Conn) verifyServerCertificate(certi
     +			DNSName:       dnsName,
      			Intermediates: x509.NewCertPool(),
      		}
    - 		for _, cert := range certs[1:] {
    + 
     
      ## src/crypto/tls/handshake_client_tls13.go ##
     @@ src/crypto/tls/handshake_client_tls13.go: import (
    @@ src/crypto/tls/handshake_messages.go: func (m *clientHelloMsg) marshal() []byte
      				// RFC 6066, Section 3
      				b.AddUint16(extensionServerName)
     @@ src/crypto/tls/handshake_messages.go: func (m *clientHelloMsg) unmarshal(data []byte) bool {
    - 		}
    + 		seenExts[extension] = true
      
      		switch extension {
     +		case extensionECH:
13:  4eb06c2026 ! 13:  f4e7958d11 Fix CIRCL integration and dependencies. (#115)
    @@ src/go/build/deps_test.go: var depsRules = `
     +
      	# crypto-aware packages
      
    - 	NET, crypto/rand, mime/quotedprintable
    + 	DEBUG, go/build, go/types, text/scanner, crypto/md5
     @@ src/go/build/deps_test.go: func findImports(pkg string) ([]string, error) {
      	if strings.HasPrefix(pkg, "golang.org") {
      		vpkg = "vendor/" + pkg
    @@ src/vendor/golang.org/x/crypto/blake2s/register.go (new)
     
      ## src/vendor/modules.txt ##
     @@
    - # golang.org/x/crypto v0.0.0-20211215153901-e495a2d5b3d3
    + # golang.org/x/crypto v0.0.0-20220516162934-403b01795ae8
      ## explicit; go 1.17
     +golang.org/x/crypto/blake2b
     +golang.org/x/crypto/blake2s
14:  3df0d7487f ! 14:  546a874b29 Add hybrid post-quantum key agreement.
    @@ src/crypto/tls/cfkem_test.go (new)
     
      ## src/crypto/tls/handshake_client.go ##
     @@ src/crypto/tls/handshake_client.go: type clientHandshakeState struct {
    - 	session      *ClientSessionState
    - }
    + 
    + var testingOnlyForceClientHelloSignatureAlgorithms []SignatureScheme
      
     -func (c *Conn) makeClientHello(minVersion uint16) (*clientHelloMsg, ecdheParameters, error) {
     +func (c *Conn) makeClientHello(minVersion uint16) (*clientHelloMsg, clientKeySharePrivate, error) {
    @@ src/crypto/tls/handshake_client.go: type clientHandshakeState struct {
      	if len(config.ServerName) == 0 && !config.InsecureSkipVerify {
      		return nil, nil, errors.New("tls: either ServerName or InsecureSkipVerify must be specified in the tls.Config")
     @@ src/crypto/tls/handshake_client.go: func (c *Conn) makeClientHello(minVersion uint16) (*clientHelloMsg, ecdheParamet
    - 		hello.supportedSignatureAlgorithms = config.supportedSignatureAlgorithms()
    + 		hello.supportedSignatureAlgorithms = testingOnlyForceClientHelloSignatureAlgorithms
      	}
      
     -	var params ecdheParameters
```
</details>

Once reviewed and verified, it will be force-pushed.

cjpatton commented:
Looks good to me, thanks @Lekensteyn!

Lekensteyn commented:
In the last push, I fixed the Go `TestStd` failure and addressed some deprecation warnings in the GitHub workflow tests:

<details>
<summary>changes since previous push: git range-diff @{u}...</summary>

```
 1:  f3b0f0f926 !  1:  6da32c3fef .github: add docker-compose.yaml and GitHub Action for CI
    @@ .github/workflows/test.yml (new)
     +  test:
     +    runs-on: ubuntu-latest
     +    steps:
    -+      - uses: actions/checkout@v2
    ++      - uses: actions/checkout@v3
     +      - name: Run tests
    -+        run: ./all.bash --no-clean
    ++        run: ./all.bash
     +        working-directory: src
     
      ## docker-compose.yaml (new) ##
 2:  89a9569c58 =  2:  af885f3146 .github: replace upstream templates and CODE_OF_CONDUCT.md
 3:  c4feb39122 =  3:  e3e940eece cmd/api, go/build: ignore CF-specific API changes
 4:  6a873c45b0 =  4:  7282b47269 circl: import crypto library CIRCL
 5:  afc5595d58 =  5:  e1aa8f69d1 crypto: add support for CIRCL signature schemes
 6:  8c79656aef =  6:  533a9a76f4 crypto/tls: add PQSignatureSchemesEnabled option
 7:  de9cdcf3c9 =  7:  a2df9ec569 crypto/tls: add CFControl parameter to Config
 8:  2814068d4e =  8:  64889b7911 crypto/tls: add Config.CFEventHandler to expose intra-handshake timing
 9:  f1691ca795 =  9:  8b9ee58155 crypto/tls: implement draft-ietf-tls-subcerts-10
10:  7fb630ea4c = 10:  2510ae5f02 crypto/tls: implement draft-ietf-tls-esni-13
11:  f4e7958d11 ! 11:  f7eb97bd4a Fix CIRCL integration and dependencies. (#115)
    @@ Commit message
         Certificates were hardcoded, now the PQ certificate is created
         as part of the test script.
     
    +     [pwu: Go 1.19: added circl to src/go/doc/comment/std.go]
    +
      ## cf-copy-from-circle.py ##
     @@ cf-copy-from-circle.py: import subprocess
      base = os.path.dirname(os.path.abspath(sys.argv[0]))
    @@ src/go/build/deps_test.go: func findImports(pkg string) ([]string, error) {
      	files, err := os.ReadDir(dir)
      	if err != nil {
     
    + ## src/go/doc/comment/std.go ##
    +@@ src/go/doc/comment/std.go: package comment
    + var stdPkgs = []string{
    + 	"bufio",
    + 	"bytes",
    ++	"circl",
    + 	"context",
    + 	"crypto",
    + 	"embed",
    +
      ## src/vendor/github.com/bwesterb/go-ristretto/.travis.yml (new) ##
     @@
     +language: go
12:  546a874b29 = 12:  6ca04bb8cc Add hybrid post-quantum key agreement.
```
</details>

If this passes, I'll complete this PR (force push to `cf`).

bwesterb commented:
I think we get around the API issue in goboring-pq by removing `VERSION`, but it's a while ago, so I'm not 100% sure.

Lekensteyn commented:
> I think we get around the API issue in goboring-pq by removing `VERSION`, but it's a while ago, so I'm not 100% sure.

We probably need the VERSION file to make things like `go version` work. The problem is that commit b7041c7ad1c9d42078cfc376320e5b307e617a80 hard-coded the `-allow_new=false` option which negates the special code we added to assume `-allow_new=true` in commit `cmd/api, go/build: ignore CF-specific API changes`.

I am exploring a couple of options to address that. I could somehow activate that `-allow_new=true` option, add a special `api/go1.*.txt` (e.g. `go1.000.txt`) file with the new symbols, or something else.

Lekensteyn commented:
After chatting with Bas, he raised the concern that we need to update this file every time there are new API changes. The test is only run at the end which result in a slower feedback cycle.

Therefore I dropped the `api/go1.000.txt` and moved the special code to disable failures on new APIs. For future reference, this was the comment that explains why the API file was named as such:
```
# This Cloudflare Go fork does not guarantee a stable API, new symbols can be
# added at any point. They must be listed here to please the "API check" test:
#   bin/go tool dist test -run api --no-rebuild
#
# Why go1.000.txt as filename? See README. Since Go 1.19 (commit b7041c7ad1c9),
# every line in go1.19.txt and next/*.txt must end with "#nnnnn". Either that,
# or pretend to be an older version so we can reuse the API check output.
```
___
In the latest push, I dropped the latest api/go1.000.txt file and changed the "cmd/api, go/build: ignore CF-specific API changes" commit message as follows:
```diff
     cmd/api, go/build: ignore CF-specific API changes

-    In order to reduce the chances of merge conflicts to the api/go1.X.txt
-    or api/next.txt files, avoid requiring new APIs to be listed in either
-    file when the `-cf` tag is present in the VERSION file.
+    The "API check" test requires new APIs to be tracked in api/go1.X.txt or
+    api/next/X.txt. Since Go 1.19 (commit b7041c7ad1c9), every line in these
+    files also need a comment with the approval issue number. To reduce
+    development friction, we disable the requirement of updating these files
+    when the `-cf` tag is present in the VERSION file.

     The circl library has a large number of packages. We do not even use all
     of it. To save us some work, just skip validating the import
```

The check was moved earlier in the file with an updated comment:
```diff
--- a/src/cmd/api/run.go
+++ b/src/cmd/api/run.go
@@ -90,4 +90,12 @@ func approvalNeeded(files []string) []string {
 // allowNew returns the -allow_new flag to use for the 'go tool api' invocation.
 func allowNew() string {
+	// The CF Go fork adds new APIs to crypto/x509 and crypto/tls, make sure
+	// that these do not fail the "API check" test. API additions could be
+	// tracked in files such as api/go1.000.txt, but since there are no API
+	// stability commitments, the extra work to maintain it is not worth it.
+	if strings.Contains(runtime.Version(), "-cf") {
+		return "-allow_new=true"
+	}
+
 	// Experiment for Go 1.19: always require api file updates.
 	return "-allow_new=false"
@@ -99,12 +107,4 @@ func allowNew() string {
 	}
 
-	// We would like to be able to add new APIs without having to update the
-	// API file since that is prone to running into merge conflicts during
-	// updates. The circl library for example has a large API surface which
-	// is not stable.
-	if strings.Contains(runtime.Version(), "-cf") {
-		return "-allow_new=true"
-	}
-
 	// See whether the api/go1.n.txt for this Go version has been created.
 	// (As of April 2021, it gets created during the release of the first Beta.)
```
